Finite strain kinematics updates on Hill material

recuero opened the pull request:
## Reason
Structural elements undergoing large displacements and/or large deformations can cause significant rotations in the material. For these reason, the anisotropic directions provided by the user may no longer be acceptable as the material deforms.

This PR also includes:
 - A fix to the creep flow rule
 - A change in the way the Hill's tensor is transformed (no need for rotation sequences). The transformation happens directly on the Voigt-notation Hill's compliance tensor with components from rotation matrices.
 - A change to add default transformation of the Hill tensor.

## Design
Account for rotation kinematics in the `HillConstants` material and update the anisotropic coefficients at each iteration. This, in the general case, involves considering the full transformation of the tensor, which can then be used in the `stressUpdate` constitutive models. 

## Impact
Anisotropic creep/plasticity simulations where large rotation occurs can be more accurate.


Closes #16016

recuero committed:
Add influence of finite strain kinematics on Hill material (#16016)

Initial testing only about Z axis confirms the parameters are being rotated
internally as per the finite strain kinematic decomposition.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/784474/) on c758516 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18453/site)

This comment will be updated on new commits.

SudiptaBiswas committed:
Alternative design for Hill tensor rotation #16016

recuero committed:
Conditional creation of material properties and debugging (#16016)

recuero committed:
Make transformation matrix non-AD initially (#16016)

recuero committed:
Allow the user to select large rotation approach to the definition of anisotropic coefficients (#16016)

recuero committed:
Update tests for flow rule fix and add large deformation rotation tests (#16016)

The flow rule was incorrect. This commit updates the gold files to reflect
correct results. In addition, a number of tests rotating a structural element
are added, such that the rotation kinematics drives the transformation of the
Hill tensor.

recuero committed:
Minor cleanup (#16016)

dschwen commented on modules/tensor_mechanics/include/materials/HillConstants.h:
```diff
@@ -58,4 +73,8 @@ class HillConstants : public ADMaterial
+
+private:
+  /// Compute Euler angles in <Z Y X> sequence from strain kinematic rotation matrix
+  std::array<Real, 3> computeZYXAngles(const RankTwoTensor & rotation_matrix);
```
We have an `EulerAngles` class that this function should be added to. And if I'm not mistaken @SudiptaBiswas did implement a rotation matrix -> euler angles routine before.

dschwen replied:
Here: https://github.com/idaholab/moose/blob/e65c36f92f31449296c2cb86a1cd463e0cb3c65f/modules/phase_field/src/userobjects/EulerAngleUpdater.C#L108-L155

recuero replied:
Looks like the angle sequence there is different.

dschwen replied:
Yes, but does it have to be? It would be less confusing if we had one single convention for Euler angles in MOOSE.

SudiptaBiswas replied:
`EulerEngleUpdater` follows ZX'Z'' transformation and it was designed for a specific application. I agree with @dschwen, we should refactor this angle extraction operation in a separate class.

recuero replied:
> Yes, but does it have to be?

With the existing code, yes, it has to because of how the transformation of the Hill tensor was done. We would need to also change that to have the same sequence as the one present in the phase field module.

recuero replied:
The modifications last pushed transform the Hill's tensor in Voigt notation using the components of the total rotation matrix, so there isn't a need to define rotation sequences. For all the anisotropic classes, the transformed matrix will be used by default. An additional option is provided to the user when using the creep class: The Hill coefficients are extracted at the beginning of the solution (`use_large_rotation = false`) and the ADHillCreep uses such coefficients by prescribing `use_transformation = false`. This is done to avoid doing unnecessary computations when transformations aren't needed, but, from the tests, the performance differences may not be very large.

SudiptaBiswas commented on modules/tensor_mechanics/src/materials/HillConstants.C:
```diff
@@ -239,4 +327,23 @@ HillConstants::rotateHillConstants(const std::vector<Real> & hill_constants_inpu
+  _hill_tensor(5, 5) = 2.0 * M;
+
+  _hill_tensor.right_multiply_transpose(_transformation_tensor);
+  _hill_tensor.left_multiply(_transformation_tensor);
```
We should either rotate the `HillConstants` or the `HillTensor` depending on the rotation angle. We should add a selection criterion here.

recuero replied:
The `stressUpdate` material would need to know what's available, constants or tensor. If the material provide both the selection can take place in the constitutive model class. If we choose to rotate only one of them here, we may need to add another user input that would need to be coordinated with, e.g., the anisotropic creep class.

@dschwen any idea to make this design cleaner?

SudiptaBiswas replied:
How about we keep both material properties but only rotate whichever is necessary? We can build the updated Hill tensor from the rotated Hill constants for the first case.

recuero replied:
I think it can be involved to decide which transformation is necessary without adding flags to coordinate with the stress update classes (creep can work with constants, but plasticity doesn't really benefit much from constants since it performs tensor operations).  We know we cannot describe large deformation through Hill constants since they cannot be extracted from a generally transformed Hill's tensor. I added an if statement to avoid performing the extraction of these components when we know it won't be needed. This can save us some time when we choose to include large deformation rotation in the anisotropy directions.

SudiptaBiswas commented on modules/tensor_mechanics/src/materials/ADHillCreepStressUpdate.C:
```diff
@@ -79,20 +85,30 @@ ADHillCreepStressUpdate::computeResidual(const ADDenseVector & /*effective_trial
+    ADDenseVector Ms(6);
+    _hill_tensor[_qp].vector_mult(Ms, stress_new);
+    qsigma_square = Ms.dot(stress_new);
+  }
```
Shouldn't similar operations be added to other Hill tensor classes?

SudiptaBiswas commented on modules/tensor_mechanics/src/materials/ADHillCreepStressUpdate.C:
```diff
@@ -30,6 +30,10 @@ ADHillCreepStressUpdate::validParams()
+  params.addParam<bool>("use_transformation",
+                        false,
+                        "Whether to employ updated Hill's tensor due to rigid body or large "
+                        "deformation kinematic rotations");
```
We probably would be able to use the optimized version for some transformation angles like 90, 180 etc. Maybe we should add a comment regarding that.

recuero committed:
Address Sudipta's review (#16016)

- Extend use of transformed Hill's tensor to HillPlasticity and HillElastoPlasticity. This classes now
  use the Hill tensor and not coefficients since they internally deal with tensor math. Gold files
  have been updated because the equivalent plastic strain slightly changes with this commit and this
  has an influence on the material time step predicted. The strain and stress remain unchanged though.
- Avoid extracting Hill's constants in the Hill material when the user flags that we are modeling
  rotation due to large deformation (i.e. Hill's constants are not sufficient, we need full transformed
  tensor). A couple of tests are deleted to account for this.
- Add comment to user input.

recuero committed:
Address Daniel's comments (#16016)

Avoid using Euler angle sequences and perform the transformation of the 6-by-6
Hill's tensor directly from the components of the rotation matrix resulting from
large deformation rotation and a possible initial orientation.

Now, we avoid extracting Hill constants when we transform the Hill's tensor. The
default value for the HillConstant material is to compute the transformation,
and the default for the creep model is to use such a transformation. To support
simple cases where changes of orientation aren't needed, the use can select
use_transformation = false in the creep model and use_large_rotation = false
in the HillConstant material.

recuero committed:
Call computeStrainFinalize no matter what (#16016)

If we don't, if the system experiences intermittently, stateful properties
such as inelastic_effective_strain won't get updated properly, which is an
issue that goes beyond reporting and could cause the appearance of a negative
time step, which will kill the simulation.

dschwen commented on modules/tensor_mechanics/src/materials/ADHillCreepStressUpdate.C:
```diff
@@ -154,23 +174,32 @@ ADHillCreepStressUpdate::computeStrainFinalize(ADRankTwoTensor & inelasticStrain
+    qsigma_square = Ms.dot(stress_dev);
+  }
+
   qsigma_square = std::sqrt(qsigma_square);
```
I'd move that square root below the conditional (with the change below)

dschwen commented on modules/tensor_mechanics/src/materials/ADHillCreepStressUpdate.C:
```diff
@@ -185,28 +214,49 @@ ADHillCreepStressUpdate::computeStrainFinalize(ADRankTwoTensor & inelasticStrain
-      prefactor / std::sqrt(2 * M) * 2.0 * M * stress(0, 2);
-  inelasticStrainIncrement(1, 2) = inelasticStrainIncrement(2, 1) =
-      prefactor / std::sqrt(2 * L) * 2.0 * L * stress(1, 2);
+  ADReal prefactor = delta_gamma / qsigma_square;
```
```suggestion
  ADReal prefactor = delta_gamma / std::sqrt(qsigma_square);
```

this seems more readable

dschwen commented on modules/tensor_mechanics/src/materials/ADHillElastoPlasticityStressUpdate.C:
```diff
@@ -54,7 +54,7 @@ ADHillElastoPlasticityStressUpdate::ADHillElastoPlasticityStressUpdate(
     _yield_stress(getParam<Real>("yield_stress")),
     _hill_constants(getMaterialPropertyByName<std::vector<Real>>(_base_name + "hill_constants")),
-    _hill_tensor(6, 6)
+    _hill_tensor(getMaterialPropertyByName<DenseMatrix<Real>>(_base_name + "hill_tensor"))
```
Just to clarify, there is no "use transformation: option here? The full tensor is always required?

dschwen replied:
I'm asking because not always  declaring the full tensor property would be a way to make MOOSE error out if inconsistent options are used.

recuero replied:
The `hill_tensor` property is now always handled as a pointer to enable conditional declaration/retrieval.

dschwen reviewed (commented):
See comments

recuero committed:
Address Daniel's comments II and some (#16016)

 - Do not retrieve properties that aren't used
 - Use conditional declaration and retrieval of the hill tensor material
   to ensure proper error when the user doesn't adequately coordinate the
   settings of the HillConstant and HillStressUpdate materials.
 - Small reorg of qsigma in ADHillCreepStressUpdate
